File: 0048-rotate-image/0048-rotate-image.ts
/**
 Do not return anything, modify matrix in-place instead.
 */
function rotate(matrix: number[][]): void {
    for(let i=0 ; i<matrix.length ; i++){
        for(let j=i+1 ; j<matrix.length ; j++){
            let temp = matrix[i][j]
            matrix[i][j] = matrix[j][i]
            matrix[j][i] = temp
        }
    }

    for(let i=0 ; i<matrix.length ; i++){
        reverse(matrix[i])
    }

};

function reverse(arr: number[]): void{
    let s = 0
    let e = arr.length-1

    while(s<e){
        let temp = arr[s]
        arr[s] = arr[e]
        arr[e] = temp

        s++
        e--
    }
}

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void reverse(vector<int> &arr){
        int i=0;
        int j=arr.size()-1;
        while(i<j){
            swap(arr[i], arr[j]);
            i++;
            j--;
        }
    }
    void rotate(vector<vector<int>>& matrix) {
        // transpose

        int n = matrix.size();
        int m = matrix[0].size();

        for(int i=0 ; i<n ; i++){
            for(int j=i+1 ; j<m ; j++){
                swap(matrix[i][j], matrix[j][i]);
            }
        }

        // reverse
        for(int i=0 ; i<n ; i++){
            reverse(matrix[i]);
        }
    }
};